JNI DETECTED ERROR IN APPLICATION: JNI GetObjectRefType called with pending exception java.lang.RuntimeException: java.lang.AssertionError: Illegal type provided

<!--
NOTE: please submit only bug reports here, any new questions or feature requests should be submitted in Discussions:
https://github.com/software-mansion/react-native-reanimated/discussions
 -->

## Description

The worklets in Reanimated v2 seems to be really unstable in Android. Every change in the worklet code crashes the app and the app needs to be restarted like 5 times (crashes 5 times). And if you have an error in the worklet (e.g. calling any standard javascript function like parseInt, Math.round etc) it crashes in the native code with always the same error. 

If I move the code from onEnd worklet to a normal function and call it with runOnJS it either does nothing or the app crashes. Is runOnJS even supposed to work?

```
 java_vm_ext.cc:578] JNI DETECTED ERROR IN APPLICATION: JNI GetObjectRefType called with pending exception java.lang.RuntimeException: java.lang.AssertionError: Illegal type provided
 java_vm_ext.cc:578]   at void com.swmansion.reanimated.AndroidErrorHandler.raise(java.lang.String) (AndroidErrorHandler.java:6)
 java_vm_ext.cc:578]   at void com.swmansion.reanimated.Scheduler.triggerUI() (Scheduler.java:-2)
 java_vm_ext.cc:578]   at void com.swmansion.reanimated.Scheduler$1.run() (Scheduler.java:24)
```

### Expected behavior

Crashes would be captured and a stack trace printed to the console. Or a proper error in the native logs. And when restarting the app after fixing the code (or not changing it at all when no bug in the code) the app would work in immediately not after 5 to 10 restarts (some caching issue?). 

### Actual behavior & steps to reproduce

Unhandled crashes and no proper recovery.

## Snack or minimal code example

I'm using the pan gesture handler worklets with Samsung Galaxy A51 5G (SM-A516B).

## Package versions

- React Native: 0.66.3
- React Native Reanimated: 2.3.0
- React Native Gesture Handler: 2.1.0
- NodeJS: 14.18.1
- Java & Gradle: OpenJDK 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing) & 7.3

## Affected platforms

- [X] Android
- [ ] iOS
- [ ] Web

> Hey, could you provide a repo to reproduce your issue? Without repro, I am not able to help 😕

Hey @piaskowyk:

I finally tried updating to `2.3.1` as a prerequisite of updating to RN 0.67 after previously being stuck on `2.2.3`, which was the last good version not to have this error. Not sure if you remember but I've been on a big rabbit hole trying to troubleshoot and create a repro for this in various threads.

Managed to create one here: https://github.com/software-mansion/react-native-reanimated/issues/2256#issuecomment-912539629

Anyways, the crash seems to be back after updating to RN 0.67-rc.6 + Reanimated 2.3.1 and forcing Hermes to version `0.10.0` via yarn `resolutions`. It's likely these steps are not necessary.

<details>
```
java_vm_ext.cc:579] JNI DETECTED ERROR IN APPLICATION: JNI GetObjectRefType called with pending exception java.lang.RuntimeException: java.lang.AssertionError: Illegal type provided
java_vm_ext.cc:579]   at void com.swmansion.reanimated.AndroidErrorHandler.raise(java.lang.String) (AndroidErrorHandler.java:6)
java_vm_ext.cc:579]   at void com.swmansion.reanimated.Scheduler.triggerUI() (Scheduler.java:-2)
java_vm_ext.cc:579]   at void com.swmansion.reanimated.Scheduler$1.run() (Scheduler.java:24)
java_vm_ext.cc:579]   at void com.swmansion.reanimated.Scheduler$2.runGuarded() (Scheduler.java:43)
java_vm_ext.cc:579]   at void com.facebook.react.bridge.GuardedRunnable.run() (GuardedRunnable.java:31)
java_vm_ext.cc:579]   at void android.os.Handler.handleCallback(android.os.Message) (Handler.java:938)
java_vm_ext.cc:579]   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:99)
java_vm_ext.cc:579]   at boolean android.os.Looper.loopOnce(android.os.Looper, long, int) (Looper.java:201)
java_vm_ext.cc:579]   at void android.os.Looper.loop() (Looper.java:288)
java_vm_ext.cc:579]   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:7798)
java_vm_ext.cc:579]   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
java_vm_ext.cc:579]   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:548)
java_vm_ext.cc:579]   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:1003)
java_vm_ext.cc:579] 
java_vm_ext.cc:579]     in call to GetObjectRefType
java_vm_ext.cc:579]     from void com.swmansion.reanimated.Scheduler.triggerUI()
```
</details>

Were you not able to reproduce the problem with the above repro?

--- a/android/src/main/java/com/swmansion/reanimated/NodesManager.java
+++ b/android/src/main/java/com/swmansion/reanimated/NodesManager.java
@@ -11,7 +11,9 @@ import com.facebook.react.bridge.ReactApplicationContext;
 import com.facebook.react.bridge.ReactContext;
 import com.facebook.react.bridge.ReadableArray;
 import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableType;
 import com.facebook.react.bridge.UiThreadUtil;
+import com.facebook.react.bridge.WritableArray;
 import com.facebook.react.bridge.WritableMap;
 import com.facebook.react.modules.core.DeviceEventManagerModule;
 import com.facebook.react.modules.core.ReactChoreographer;
@@ -612,6 +614,42 @@ public class NodesManager implements EventDispatcherListener {
     return result;
   }
 
+  private static WritableMap copyReadableMap(ReadableMap map) {
+    WritableMap copy = Arguments.createMap();
+    copy.merge(map);
+    return copy;
+  }
+
+  private static WritableArray copyReadableArray(ReadableArray array) {
+    WritableArray copy = Arguments.createArray();
+    for (int i = 0; i < array.size(); i++) {
+      ReadableType type = array.getType(i);
+      switch (type) {
+        case Boolean:
+          copy.pushBoolean(array.getBoolean(i));
+          break;
+        case String:
+          copy.pushString(array.getString(i));
+          break;
+        case Null:
+          copy.pushNull();
+          break;
+        case Number:
+          copy.pushDouble(array.getDouble(i));
+          break;
+        case Map:
+          copy.pushMap(copyReadableMap(array.getMap(i)));
+          break;
+        case Array:
+          copy.pushArray(copyReadableArray(array.getArray(i)));
+          break;
+        default:
+          throw new IllegalStateException("Unknown type of ReadableArray");
+      }
+    }
+    return copy;
+  }
+
   private static void addProp(WritableMap propMap, String key, Object value) {
     if (value == null) {
       propMap.putNull(key);
@@ -626,9 +664,17 @@ public class NodesManager implements EventDispatcherListener {
     } else if (value instanceof String) {
       propMap.putString(key, (String) value);
     } else if (value instanceof ReadableArray) {
-      propMap.putArray(key, (ReadableArray) value);
+      if (!(value instanceof WritableArray)) {
+        propMap.putArray(key, copyReadableArray((ReadableArray) value));
+      } else {
+        propMap.putArray(key, (ReadableArray) value);
+      }
     } else if (value instanceof ReadableMap) {
-      propMap.putMap(key, (ReadableMap) value);
+      if (!(value instanceof WritableMap)) {
+        propMap.putMap(key, copyReadableMap((ReadableMap) value));
+      } else {
+        propMap.putMap(key, (ReadableMap) value);
+      }
     } else {
       throw new IllegalStateException("Unknown type of animated value");
     }

